tough/src/http.rs

```diff
@@ -125,93 +132,224 @@ pub struct HttpTransport {
 }
 
 /// Implement the `tough` `Transport` trait for `HttpRetryTransport`
+#[async_trait]
 impl Transport for HttpTransport {
-    /// Send a GET request to the URL. Request will be retried per the `ClientSettings`. The
-    /// returned `RetryRead` will also retry as necessary per the `ClientSettings`.
-    fn fetch(&self, url: Url) -> Result<Box<dyn Read + Send>, TransportError> {
-        let mut r = RetryState::new(self.settings.initial_backoff);
-        Ok(Box::new(
-            fetch_with_retries(&mut r, &self.settings, &url)
-                .map_err(|e| TransportError::from((url, e)))?,
-        ))
+    /// Send a GET request to the URL. The returned `TransportStream` will retry as necessary per
+    /// the `ClientSettings`.
+    async fn fetch(&self, url: Url) -> Result<TransportStream, TransportError> {
+        let r = RetryState::new(self.settings.initial_backoff);
+        Ok(fetch_with_retries(r, &self.settings, &url).boxed())
+    }
+}
+
+enum RequestState {
+    /// A response is streaming.
+    Streaming(BoxStream<'static, reqwest::Result<bytes::Bytes>>),
+    /// A request is pending.
+    Pending(BoxFuture<'static, reqwest::Result<reqwest::Response>>),
+    /// No ongoing request.
+    None,
+}
+
+impl std::fmt::Debug for RequestState {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            RequestState::Streaming(_) => f.write_str("Streaming"),
+            RequestState::Pending(_) => f.write_str("Executing"),
+            RequestState::None => f.write_str("None"),
+        }
     }
 }
 
-/// This serves as a `Read`, but carries with it the necessary information to do retries.
 #[derive(Debug)]
-pub struct RetryRead {
+pub(crate) struct RetryStream {
```

Comment:
Does this need to be `pub(crate)`? Not sure why `RetryRead` was `pub` before.

Reply:
I will check this after merging https://github.com/awslabs/tough/issues/698